### PR TITLE
 edta draft pr

### PR DESCRIPTION
**Pull request by tcollins2011**
Creation of EDTA tool using stable docker container 2.0.1. Test data is over 1mb as a toy genome is needed for the testing to run. 

Project Home Page: https://github.com/oushujun/EDTA/blob/master/README.md

### CODE REVIEW

**Comment by mschatz** [comment]
I'd appreciate it if we could prioritize the review for this. We are hoping to submit a paper about this very soon and it would really strengthen the results to have this included. @bgruening 

Thanks!
Mike

**Comment by bgruening** [comment]
Sorry, I thought this PR is not ready yet. Its failing tests with:

> /tmp/tmpzix2swzj/job_working_directory/000/21/tool_script.sh: 9: /tmp/tmpzix2swzj/job_working_directory/000/21/tool_script.sh: Syntax error: end of file unexpected

Which looks like a serious issue.

File: tools/edta/edta.xml [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+        Whole-genome de-novo TE annotation 
+    </description>
+    <macros>
+        <token name="@WRAPPER_VERSION@">2.0.0</token>
+        <token name="@VERSION_SUFFIX@">galaxy0</token>
+        <import>edta_macros.xml</import>
+    </macros>
+    <expand macro="bio_tools"></expand>
+    <requirements>
+        <container type="docker">oushujun/edta:@WRAPPER_VERSION@</container>
```
Comment (bgruening) [inline tools/edta/edta.xml:12]: @tcollins2011 the conda package can not be used? Sorry, I do not remember why this was a problem :(

File: tools/edta/edta.xml [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+                --cds '${s.cds}'
+            #end if 
+            #if $s.curatedlib:
+                --curatedlib '${s.curatedlib}'
+            #end if
+            #if $s.sensitive:
+                --sensitive 1
+            #end if
+            #if $s.mutation_rate:
+                --u "${s.mutation_rate}"
```
Comment (bgruening) [inline tools/edta/edta.xml:71]: single quotes

File: tools/edta/edta.xml [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+            #end if
+            #if $s.exclude:
+                --exclude '${s.exclude}'
+            #end if
+            #silent$cds_list.append(str($i) + '_input.fa')
+            &&  mv ./${i}_input.fa.mod.EDTA.anno/*.mod.out .
+            &&      
+        #end for
+        #for $i in $cds_list
+            bash $__tool_directory__/filter_out_single_copies.sh $i  &&
```
Comment (bgruening) [inline tools/edta/edta.xml:84]: single-quotes

File: tools/edta/edta_macros.xml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            <option value="all">All</option>
+            <option value="filter">Filter</option>
+            <option value="final">Final</option>
+            <option value="anno">Anno</option>
+        </param>
+        <param name="cds" argument="--cds" type="data" format="fasta" label="Coding Sequence File" optional='true' help="Provide a FASTA file containing the coding sequence (no introns, UTRs, nor TEs) of this genome or its close relative."/> 
+        <param name="curatedlib" argument="--curatedlib" type="data" format="fasta" label="Curated Library File" optional='true' help="Provided a curated library to keep consistant naming and classification for known TEs. All TEs in this file will be trusted 100%, so please ONLY provide MANUALLY CURATED ones here."/>
+        <param name="sensitive" argument="--sensitive" type="boolean"  truevalue="1" falsevalue="0"  label="Use RepeatModeler to identify remaining TEs" help="Use RepeatModeler to identify remaining TEs This step is very slow and MAY help to recover some TEs."/>
```
Comment (bgruening) [inline tools/edta/edta_macros.xml:23]: when you use "argument" you do not need to use name in this case as the "name" is generated out of argument

Also a nice trick is to use `truevalue="--sensitive 1"` this also simplifies a lot your command section

File: tools/edta/edta_macros.xml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        <param name="mutation_rate" argument="--u" type="float" optional="True" label="Neutral Mutation Rate" help="Neutral mutation rate to calculate the age of intact LTR elements. Default: 1.3e-8 (per bp per year, from rice)"></param>
+        <conditional name='anno_select'>
+            <param name="anno" type="select" label="Would you like to perform whole-genome TE annotation after TE library construction?" argument='--a' help='Perform or not perform whole-genome TE annotation after TE library construction.'>
+                <option value="yes">Yes</option>
+                <option value="no">No</option>
+            </param>
+            <when value="yes">
+                <param name="evaluate" argument="--evaluate" type="boolean" truevalue="1" falsevalue="0" label="Evaluate" help="Evaluate (1) classification consistency of the TE annotation " />
```
Comment (bgruening) [inline tools/edta/edta_macros.xml:31]: remove name

File: tools/edta/edta.xml [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+        #end for
+        #if $function_select.known_te:
+            bash $__tool_directory__/make_pan_library.sh '${known_te}' 
+        #else:
+            bash $__tool_directory__/make_pan_library.sh
+        #end if 
+        #for $i in $cds_list
+            bash $__tool_directory__/annotate_pan_genome.sh $i  &&
+        #end for
+        #for $i, $s in enumerate($function_select.te_library)
```
Comment (bgruening) [inline tools/edta/edta.xml:94]: not sure why those symlinks are needed, but if the command ends with "&&" this will result in an invalid command

**Review by bgruening — COMMENTED** [review]
Some comments inline

File: tools/edta/.shed.yml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+categories:
+    - Variant Analysis
+description: |
+    The EDTA package was designed to filter out false discoveries in raw TE candidates and generate a high-quality non-redundant TE library for whole-genome TE annotations. 
+long_description: |
+    The EDTA package was designed to filter out false discoveries in raw TE candidates and generate a high-quality non-redundant TE library for whole-genome TE annotations. Selection of initial search programs were based on benckmarkings on the annotation performance using a manually curated TE library in the rice genome.
+name: edta
+owner: iuc
```
Comment (bgruening) [inline tools/edta/.shed.yml:8]: you are not on IUC here, so the owner would be "bgruening"

Comment (mschatz) [reply tools/edta/.shed.yml:8]: Thanks for all your feedback @bgruening!

File: tools/edta/edta_macros.xml (hunk 4 above)
Comment (bgruening) [reply tools/edta/edta_macros.xml:23]: ping

File: tools/edta/edta_macros.xml (hunk 5 above)
Comment (bgruening) [reply tools/edta/edta_macros.xml:31]: ping

File: tools/edta/.shed.yml (hunk 7 above)
Comment (oushujun) [reply tools/edta/.shed.yml]: Hi @bgruening, the checks have passed. Can you help to merge to the main branch? Thanks!

Comment (bgruening) [reply tools/edta/.shed.yml]: @tcollins2011 I longed you yesterday about the open comments above. Do you want to fix them as well?

**Comment by mschatz** [comment]
Woohoo!

**Comment by bgruening** [comment]
@tcollins2011 thanks! There are still a few open comments. Let me know if you want to work on them or if I should do.

**Comment by oushujun** [comment]
@bgruening Hi Bjorn, can you please help to fix the remaining issues and merge? We want to submit it on Monday. Thank you.

**Comment by bgruening** [comment]
@oushujun I think I have it running. But I found that this wrapper is using __not__ the most recent version. Is that intentional?

**Comment by oushujun** [comment]
Thank you for wrapping it up quickly. I just updated it last week when the
development of the wrapper was started a couple of weeks ago. The update
was to add the panEDTA script to run genomes sequentially, a prototype
version of galaxy panEDTA. If you could use the laterest version that would
be great, but the previous version also works fine.

**Comment by bgruening** [comment]
Ok, I will try. Thanks!

**Comment by oushujun** [comment]
@bgruening the new version is not successful. Maybe we can stick with the previous version at the moment and debug the new version later.

**Comment by bgruening** [comment]
I fixed the Docker image instead, took me a bit, but now its working with the latest version.

**Comment by oushujun** [comment]
Great! Thank you so much!

**Comment by oushujun** [comment]
@bgruening can you put the panEDTA public link here?

**Comment by bgruening** [comment]
What do you mean, public link for what? The new Docker container, TS repo or Galaxy tool. If the latter I need to install this tool. I can do so in a few hours.

**Comment by oushujun** [comment]
Sorry I mean the galaxy tool. Can people upload their genomes on a webpage
and use the cloud resource to obtain pan genome annotations? I need a link
to direct users to use or download this galaxy panEDTA. Thanks!

**Comment by bgruening** [comment]
@oushujun you can use this link here for the Galaxy tool on the European Galaxy server: https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/bgruening/edta/edta/2.1.0+galaxy0

or a shorter version https://usegalaxy.eu/root?tool_id=edta

Here are the containers: https://quay.io/repository/biocontainers/edta?tab=tags&tag=latest
Singularity images are here: https://depot.galaxyproject.org/singularity/

ToolShed repo is here: https://toolshed.g2.bx.psu.edu/view/bgruening/edta/

Hope that helps!

**Comment by mschatz** [comment]
Awesome, awesome, awesome! What are the next steps to get this into
usegalaxy.org?

**Comment by bgruening** [comment]
I think you need to add the tool to https://github.com/galaxyproject/usegalaxy-tools

**Comment by mschatz** [comment]
@tcollins2011 can you help with this? Thanks! 

Mike

**Pull request merged**